[CORL-2299] Iframeless (Shadow DOM) Epic

alinhle opened the pull request:
## What does this PR do?
- Take Coral out of the iframe

## API Deprecation
- `bodyClassName` is now deprecated. Use: `containerClassName`

## Breaking Changes 
- Raised min browser requirement, needs Shadow DOM Support, dropped support for IE11
- Font definition needs to be in a separate file (Custom Font CSS) and is configurable in admin
- `body {}`, `html {}`, `:root {}` in Custom CSS must be replaced with `#coral {}`.

## Conflict
- This PR removes feature: https://github.com/coralproject/talk/pull/3792.
  - Must be reimplemented in an iframeless way.

## Deliverables
- [x] First working version of Coral Stream outside of iframe
- [x] Bring back functionality that depended on iframe
  - [x] iframeless local, session, indexstorage
  - [x] Login / Logout
  - [x] events
  - [x] Localization
  - [x] StaticConfig
  - [x] ExternalConfig
  - [x] global CSS
  - [x] registerClickFarAway
  - [x] AMP
  - [x] Modal  / Toolbar
  - [x] Refresh Access Token
  - [x] liveCommentCount
  - [x] setCommentID
  - [x] AutoRender
  - [x] BodyClassName
  - [x] Set public path / static url
  - [x] Enable restricted Cors
  - [x] CSS Support
  - [x] Async Chunk CSS Support
  - [x] Custom CSS Support
  - [x] Custom Font Support
  - [x] Find solutions for media queries
  - [x] Error Reporter
  - [x] Isolate MobileToolbar within its own shadow dom
- [x] Expose bootstrap data at `/embed/bootstrap` 
- [x] Utilize client mounted routes in development instead of separate web dev server html files
- [x] Reduce global polluting babel Polyfills
- [x] Integrate Shadow-Dom
- [x] Integrate emotion into Shadow-Dom because of @giphy/react-components
- [x] Prevent flash of unstyled elements
- [x] Smarter preloading
- [x] Allow disabling Coral typography
- [x] Stream not as a separate library?
- [x] Deprecate/Rename BodyClassName
- [x] ~~Allow disabling individual polyfills~~
- [x] Set minimum browser requirements
- [x] Remove IE11 baggage
- [x] ~~IE11 Support~~
- [x] IframeEncapsulation
- [x] Frame bundle
- [x] Webpack isolation
- [x] Support custom Websocket ports during dev
## TODO
- [x] Resolve site
- [x] Refactor: Replace `CoralContext.renderWindow` with `CoralContext.window` (Add to backlog)
- [x] Make eventemitter universal that replaces previous pym- and postmessages between iframe and embed
- [x] Reduce global namespace pollution from polyfills
- [x] Remove global CSS
- [x] Replace StaticConfig
- [x] Replace @media queries previously relying on iframe width
- [x] Replace code that depends on `areWeInIframe()`
- [x] Replace localStorage with iframeless version
- [x] Replace sessionStorage with iframeless version
- [x] Replace indexedDBStorage with iframeless version
- [x] Replace postMessage that was used to communicate between iframes
- [x] Replace `registerClickFarAway`
- [x] Replace `pym` except on external medias
- [x] Replace `withAutoHeight` decorator
	- Support amp dynamic height
- [x] Replace `withClickEvent` decorator
- [x] Replace `withConfig` decorator
- [x] Replace `withEventEmitter` decorator
- [x] Replace `withIndexedDBStorage` decorator
- [x] Replace `withKeyPressEvent` decorator
- [x] Remove `withIOSSafariWidthWorkaround` decorator
- [x] Replace `withLiveCommentCount` decorator
- [x] Replace `withPostMessageStorage` decorator
- [x] Replace `withRefreshAccessToken` decorator
- [x] Replace `withRenderTargets` 
- [x] Remove RenderTarget functionality
- [x] Replace `withSetCommentID` decorator
- [x] Replace `OnPymLogin` listener
- [x] Replace `OnPymLogout` listener
- [x] Replace `OnEvents` listener
  - Support rudderstack events propagation
- [x] Replace `onPymSetCommentID`
- [x] Replace pym refresh token getter in `RefreshTokenHandler`
- [x] MobileToolbar / Modals / Portals must create a derived version of the ShadowRoot that also syncs CSS files into the shadow tree. MobileToolbar should be rendered using a Portal to make it independent of stacking contexes and put into its own ShadowRoot.
- [x] Merge with develop

alinhle committed:
wip

alinhle committed:
feat: mvp coral outside of iframe

alinhle committed:
wip StreamEmbed

alinhle committed:
wip: bring back more functionality

alinhle committed:
wip: Furthe bring back functionality

alinhle committed:
wip: reduce global polyfills, temporary public path, ie11 support, cors

alinhle committed:
revert removing customCSS functionality

alinhle committed:
wip: integrate shadow dom and css

alinhle committed:
wip: inject additional link tags from async chunks

alinhle committed:
wip: fix changeLocale, integrate client mounted routes into dev, expose manifest/staticConfig data

alinhle committed:
wip: bring back staticConfig, localization, custom css, static uri, prevent unstyled content flashing

alinhle committed:
wip: apply bodyClassName inside shadow dom, improve unstyled loading

alinhle committed:
wip: bugfixes, rm pym, fix tests, use staticconfig

alinhle committed:
wip: Fix ClickOutside, fix dev env, fix shadow root error, improve dev experience, allow custom fonts and disabling typography

alinhle committed:
feat: retry loading /embed/bootstrap

alinhle committed:
chore: deprecate bodyClassName for containerClassName

alinhle committed:
fix: tests

alinhle committed:
feat: intelligently preload assets when nearing embed

alinhle committed:
chore: remove support for ie11, adjust supported browsers

alinhle committed:
feat: allow disable individual polyfills

alinhle committed:
fix: tests

alinhle committed:
chore: stream not as a separte build

alinhle committed:
fix: timestamping

alinhle committed:
fix: manifest with missing integrity

alinhle committed:
chore: use same css extract mechanism for prod and dev

alinhle committed:
fix: remove polyfill config, type CoralStream, fix snapshots

alinhle committed:
fix webpack public path, static uri, <Popup />, cors issues, rename #coral-app-container to #coral

yogeshbeniwal commented:
@cvle Will this require code change in WordPress plugin also?

alinhle committed:
feat: MatchMedia now matches width of Shadow DOM, support CSS breakpoint classes

alinhle commented:
@yogeshbeniwal The api has no breaking changes and should work without changes to the Wordpress plugin.

alinhle committed:
feat: put MobileToolbar in its own derived Shadow DOM

alinhle committed:
feat: add deprecation message for bodyClassName

alinhle committed:
Merge branch 'develop' into feat/CORL-2299-iframeless-epic

alinhle committed:
fix: call onload on webpack assets

alinhle committed:
chore: upgrade webpack-assets-manifest

alinhle committed:
chore+fix: rm accidentally commited files, enable events logging per default during dev

alinhle committed:
chore: adapt browser versions

alinhle committed:
chore: refactor constants AUTORENDER_ROOT_MARGIN and PRELOAD_ROOT_MARGIN

alinhle committed:
fix: correct breakpoints for derived shadow doms that include modals

alinhle committed:
Merge branch 'develop' into feat/CORL-2299-iframeless-epic

marcushaddon commented:
Still early in testing this, but I'm unable to delete pasted text in the comment form (even after refreshing and seeing the persisted text).

kabeaty commented on src/core/client/embed/StreamEmbed.ts:
```diff
@@ -1,203 +1,318 @@
+
+  /** This is called when BootstrapConfig ist Loaded */
+  private onBootstrapConfigLoad(bootstrapConfig: EmbedBootstrapConfig) {
+    this.boostrapConfig = bootstrapConfig;
```
typo, `boostrapConfig` --> `bootstrapConfig`

kabeaty commented on src/core/client/framework/hooks/useModerationLink.ts:
```diff
@@ -0,0 +1,13 @@
+import { useCoralContext } from "coral-framework/lib/bootstrap";
+
+/**
+ * usePrevious is a react hook that will return the
```
Can you please update this comment to match the `useModerationLink` function?

kabeaty commented on src/core/client/framework/lib/errors/reporter/createReporter.ts:
```diff
@@ -19,41 +24,28 @@ function createDevelopmentReporter() {
 
 function createReporter(
   window: Window,
```
It looks like we are no longer using `window` here. Can we remove?

kabeaty commented on src/core/client/stream/tabs/Profile/ProfileQuery.tsx:
```diff
@@ -21,13 +20,9 @@ const loadProfileContainer = () =>
 
 // (cvle) For some reason without `setTimeout` this request will block other requests.
-const preloadAndPolyfill = once((window: Window) =>
+const proload = once((window: Window) =>
```
Do we still need `window` here? Looks like it isn't used now. Also, typo of `proload` --> `preload`

kabeaty commented on src/core/client/stream/tabs/Discussions/DiscussionsQuery.tsx:
```diff
@@ -21,13 +20,9 @@ const loadDiscussionsContainer = () =>
 
 // (cvle) For some reason without `setTimeout` this request will block other requests.
-const preloadAndPolyfill = once((window: Window) =>
+const preload = once((window: Window) =>
```
Do we still need `window` here? Looks like it isn't used now.

kabeaty commented on src/core/client/stream/tabs/Configure/ConfigureQuery.tsx:
```diff
@@ -20,13 +19,9 @@ const loadConfigureContainer = () =>
 
 // (cvle) For some reason without `setTimeout` this request will block other requests.
-const preloadAndPolyfill = once((window: Window) =>
+const preload = once((window: Window) =>
```
Do we still need `window` here? Looks like it isn't used now.

kabeaty commented on src/core/client/ui/components/v2/MatchMedia/MatchMedia.tsx:
```diff
@@ -46,6 +58,86 @@ interface Props {
+  }, [shadowRoot, minWidth, maxWidth]);
+
+  const minWidthChanged = prevMinWidth !== null && prevMinWidth !== minWidth;
+  const maxWidthChanged = prevMaxWidth !== null && prevMaxWidth !== minWidth;
```
Should this be checking for `prevMaxWidth !== maxWidth` instead?

kabeaty commented on src/core/client/ui/shadow/ReactShadowRoot.tsx:
```diff
@@ -0,0 +1,122 @@
+  cssAssets?: CSSAsset[];
+  /** CSS assets to be added at the end */
+  customCSSAssets?: CSSAsset[];
+  style?: React.CSSProperties;
```
I think it could be nice to add a comment here for `style` too about how it's used for hiding unstyled content before css is loaded.

kabeaty commented on src/core/client/ui/shadow/ReactShadowRoot.tsx:
```diff
@@ -0,0 +1,122 @@
+ * it derives all props that are set on `ReactShadowRoot` essentially giving
+ * you the same shadow dom environment with everything setup like CSS assets loading.
+ */
+export const ReactShadowRootDerived: FunctionComponent<DerivedProps> = (
```
I think this name works fine, but other ideas would be `ReactShadowRootDerivedChild` (longer but more explicit about how must be a descendant of main shadow root element) or just `ReactShadowRootChild` (although this might lose some of the clarity that it inherits from `ReactShadowRoot`).

alinhle replied:
There is no longer a `ReactShadowRootDerived` component and the ShadowRoot has been refactored into a higher level `Encapsulation` concept as we now also have an `IframeEncapsulation` component.

kabeaty commented on src/core/server/app/helpers/manifestLoader.ts:
```diff
@@ -0,0 +1,168 @@
+/**
+ * Sometimes webpack produces an invalid manifest with empty entrypoints.
+ * We retry whenever we encounter this and usually webpack fixes this.
+ * This is a hardlimit on how many retires we should do.
```
typo, `retires` --> `retries`

kabeaty commented on src/core/server/app/middleware/graphql/apolloServer.ts:
```diff
@@ -132,14 +133,17 @@ export const apolloGraphQLMiddleware = ({
+  // Check request coming from known origin.
+  const corsOptionsDelegate = createCorsOptionsDelegate(options.mongo);
+
+  // Get the GraphQL miqddleware.
```
typo, `miqddleware` --> `middleware`

kabeaty commented on src/locales/en-US/admin.ftl:
```diff
@@ -857,6 +857,11 @@ configure-wordList-suspect-wordListDetailInstructions =
 configure-advanced-customCSS-override =
   URL of a CSS stylesheet that will override default Embed Stream styles.
+configure-advanced-customCSS-stylesheetURL = Custom CSS Stylesheet URL
+configure-advanced-customCSS-fontsStylesheetURL = Custom CSS Style Sheet URL for Font Faces
```
Can we change `Style Sheet` --> `Stylesheet` for consistency here? Also, for `configure-advanced-customCSS-containsFontFace` below, I think `stylesheets` should just be singular `stylesheet` (either capitalized like the others, or all lowercased).

kabeaty commented on docs/docs/cms.md:
```diff
@@ -37,8 +37,9 @@ Options object for `createStreamEmbed`:
 - `accessToken` - [string] _(optional)_: Access token to log in a user via [SSO](/sso).
 - `refreshAccessToken` - [function] _(optional)_: Callback to obtain a new access token when the current one has expired. A parameter `nextAccessToken` is passed as the first argument that should be called with the next access token.
-- `bodyClassName` - [string] _(optional)_: HTML class name to add to the body of the stream embed iframe for CSS targeting.
+- `containerClassName` - [string] _(optional)_: HTML class name to add to the container of the stream embed iframe for CSS targeting.
```
Should this comment just say `container of the stream embed for CSS...` without `iframe`?

kabeaty commented on docs/docs/css.md:
```diff
@@ -35,11 +35,11 @@ You can also navigate to https://github.com/coralproject/talk/blob/main/src/core
 ### Custom body class for theming
 
-You can set the class name of the `<body>` tag inside the embed by using the `bodyClassName` parameter when calling `Coral.createStreamEmbed`:
+You can set the class name of the `<body>` tag inside the embed by using the `containerClassName` parameter when calling `Coral.createStreamEmbed`:
```
Should this doc and its header be updated to something more like `You can set the class name of the container inside of the embed by using...` and `Custom container class for theming` without a reference to the `<body>` tag?

kabeaty commented on src/core/build/createWebpackConfig.ts:
```diff
@@ -471,15 +476,56 @@ export default function createWebpackConfig(
+                  loader: require.resolve("style-loader"),
+                  options: {
+                    injectType: "linkTag",
+                    /* , insert: insertLinkTag */
```
Why is this commented out?

nick-funk replied:
They moved it up into the `MiniCssExtractPlugin` 😃

alinhle replied:
Will be removed

kabeaty commented on src/core/client/embed/Coral.ts:
```diff
@@ -30,35 +30,51 @@ export interface Config {
   enableDeprecatedEvents?: boolean;
 
   /** Allow setting className of body tag inside iframe */
```
Can this comment be updated to match `containerClassName`?

kabeaty commented on src/core/client/embed/test/basic.spec.ts:
```diff
@@ -1,56 +1,204 @@
+      server.restore();
+    }
   });
   it("should remove iframe", () => {
```
Can the name of this test be updated?

kabeaty commented on src/core/client/embed/test/basic.spec.ts:
```diff
@@ -1,56 +1,204 @@
-    CoralEmbedStream.render();
-    expect(container.innerHTML).toMatchSnapshot();
-    document.head.removeChild(link);
-    expect(console.warn).not.toHaveBeenCalled();
```
Should this check still be included in this test? Looks like it is calling a warning about canonical link here when previously we expected it not to?

alinhle replied:
readded!

kabeaty commented on src/core/client/framework/components/loadables/MarkdownEditor/MarkdownEditor.css:
```diff
@@ -561,11 +561,14 @@ $fullscreenZIndex: 10;
     }
   }
+
+  /* @cvle: This part does not seem to be used? */
```
Do you think we can safely remove this then, or should we test further first to confirm?

alinhle replied:
I think it's safe

kabeaty commented on src/core/client/stream/common/KeyboardShortcuts/KeyboardShortcuts.tsx:
```diff
@@ -529,10 +518,13 @@ const KeyboardShortcuts: FunctionComponent<Props> = ({ loggedIn }) => {
         // in instead of waiting for polling to update it
-        pym.sendHeight();
+
+        // TODO: (@cvle) pym no longer exist, we usually don't need this. However we might need a mechanism like this for amp?
```
Do you recommend we just do some amp testing to confirm this?

alinhle replied:
I readded it.

kabeaty commented on src/core/client/stream/App/listeners/OnEmbedSetCommentID.ts:
```diff
@@ -1,26 +1,24 @@
+        const id = value || null;
         if (s.get(LOCAL_ID)!.getValue("commentID") !== id) {
           s.get(LOCAL_ID)!.setValue(id, "commentID");
 
```
Do we still need the below, or can it be removed?

alinhle replied:
Should be removed

kabeaty commented on src/core/client/embed/StreamEmbed.ts:
```diff
@@ -1,203 +1,318 @@
         // user to the embed.
         this.render();
       } else {
+        // Preload assets when nearing embed.
```
🎉

kabeaty reviewed (changes requested):
Such exciting changes! Looking good overall to me. I left some comments and questions throughout.   

Through manual testing, I also noticed some things that I added to the QA doc. 
Outside of those, I’m also seeing locally: 
- some warnings in the console about “couldn’t fulfill desired order of chunk groups”. Do you know what’s causing this?
- On the “Story with button” page, clicking “Show comments” doesn’t load the comments. I’m seeing `http://undefined/embed/bootstrap` in the console on this page, so looks like it’s not getting the rootURL from config.  
- I also did some searches for “pym” and “iframe” in the code. I’m still seeing a few mentions of both in the code that I think can be removed/updated. For example, in `retrieveSiteFromQuery`, `CallOut.tsx`, `getAriaPoliteMacOSWorkaround`, and a few more files. Also some docs such as in `seo.md`.

alinhle committed:
fix: pass static config

alinhle committed:
fix: populate static config and fix keyboard shortcuts

alinhle committed:
fix: set rootURL

nick-funk commented on package.json:
```diff
@@ -466,12 +468,11 @@
-    "Edge >= 17",
-    "Firefox >= 68",
-    "Chrome >= 49",
+    "iOS >= 10",
```
Just a suggestion, but as of XCode 12.1, Apple recommends apps to support iOS 12 - 15 presently. Are there clients we know of that are using iOS 10, 11?

alinhle replied:
iOS 10 and iOS11 work just fine without special treatment, do we currently gain anything from removing support of the two versions?

nick-funk replied:
Ah, just trying to cut down our testing. The less devices we support, the less we have to test. No worries, we can leave it as-is.

alinhle committed:
fix: immediately update amp height

alinhle committed:
fix: test

alinhle committed:
fix: add comment

alinhle committed:
fix: apply review

alinhle committed:
fix: tests from review

nick-funk commented on src/core/common/utils/getHost.ts:
```diff
@@ -0,0 +1,4 @@
+export default function getHost(url: string) {
+  const split = url.split("/");
+  return split[2];
+}
```
can we do something like the following here?

```
const path = new URL(url);
const host = path.hostname;
```

alinhle replied:
Hmm I guess, though this implementation can be used server side as well, so as long as the test passes, I think it's fine?

nick-funk replied:
You can use `URL` class server side as well.

I ran the following in node server side:
```
const rawURL = "https://www.theverge.com/2022/2/2/22914970/facebook-app-loses-daily-users-first-time-earnings";

const url = new URL(rawURL);
console.log("url based hostname:", url.hostname);

const split = rawURL.split("/");
console.log("split based hostname:", split[2]);
```

Got this result:
```
url based hostname: www.theverge.com
split based hostname: www.theverge.com
```

The only reason I suggest it is that URL will validate the link with all edge cases. For instance, web browsers will accept `http:/` as well as `http://`. The `new URL(...)` library does this as well, but the split based method will fail and return 2022 for the above example `rawURL`.

Also, `URL` will throw other errors if the URL is malformed in other weird ways and give an informative message.

alinhle replied:
Ah didn't know it works serverside. I guess I can use  `new URL(url).host` directly then, thanks!

alinhle replied:
Actually no, it doesn't support omitting the protocol like `//domain.com` which browsers do accept and is standardized. `coral-party` repository e.g. is depending on it.

nick-funk replied:
Browsers only accept `//domain.com` to point to `file:////domain.com` though? We use that in `coral-party`?

nick-funk replied:
Anyways, this is fine for now. Not holding a review over a two line change.

nick-funk reviewed (changes requested):
This is an incredible amount of work, thank you so much for all your efforts @cvle !

I am only putting `Request changes` because I found some bugs. I pulled recent changes and the RTE text input still seems a bit weird and after I post a comment, the RTE freezes up and I can't type into it again.

The second issue I ran into is that the live updates websockets can't seem to connect. I'm digging into this to see if I can find out why.

![image](https://user-images.githubusercontent.com/5751504/152008645-3b96570b-5f8f-4760-80f4-76ca532b2eef.png)

I'm sure we'll get these issues resolved, amazing work and thank you again for all your effort into setting this up for all of us!

alinhle committed:
fix: graphql cors

alinhle committed:
feat: RTE in iframe mvp

alinhle committed:
refactor: Introduce EncapsulationContext

alinhle committed:
refactor+feat: refactor iframe+shadowroot into common encapsulation concept + improvements, CoralContainer with breakpoints now ubiquitous

alinhle committed:
chore: improve comment

alinhle committed:
fix: improve getCurrentScriptOrigin

alinhle committed:
fix: support older contentRec ResizeObserverEntry Prop

alinhle committed:
fix: cyclic dep

alinhle committed:
fix: CSSAssets onload should only called once, fix iframe margin

alinhle committed:
fix: set correct port in webpack dev server

alinhle committed:
fix: allow cors port websockets in development

alinhle committed:
feat+chore: improve iframe handling, asset loading, and error display

alinhle committed:
feat: replace pym with frame bundle

alinhle committed:
fix: no need to split frame bundle into chunks

alinhle committed:
Merge branch 'develop' into feat/CORL-2299-iframeless-epic

alinhle committed:
fix: remove iframe scrolling

alinhle committed:
fix: assign unique jsonpFunction to our webpack bundles

alinhle committed:
chore: removing some remaining pym and iframe references

alinhle committed:
chore: dont crash server so quickly

alinhle commented:
> This is an incredible amount of work, thank you so much for all your efforts @cvle !
> 
> I am only putting `Request changes` because I found some bugs. I pulled recent changes and the RTE text input still seems a bit weird and after I post a comment, the RTE freezes up and I can't type into it again.
> 
> The second issue I ran into is that the live updates websockets can't seem to connect. I'm digging into this to see if I can find out why.
> 
> ![image](https://user-images.githubusercontent.com/5751504/152008645-3b96570b-5f8f-4760-80f4-76ca532b2eef.png)
> 
> I'm sure we'll get these issues resolved, amazing work and thank you again for all your effort into setting this up for all of us!

This should be addressed now at least during development. Using custom ports confuses some code we introduced for the nginx proxy support.

![image](https://user-images.githubusercontent.com/14221600/152241458-043308ff-e6e5-44e2-beea-b2644dec60e3.png)

alinhle commented:
> Such exciting changes! Looking good overall to me. I left some comments and questions throughout.   
> 
> Through manual testing, I also noticed some things that I added to the QA doc. Outside of those, I’m also seeing locally: 
> 
> * some warnings in the console about “couldn’t fulfill desired order of chunk groups”. Do you know what’s causing this?

This means that in one bundle/chunk we asked for one ordering while in another we asked for the opposite order to load the CSS. This should do no harm though, you can try to track down this and fix the ordering.

> * On the “Story with button” page, clicking “Show comments” doesn’t load the comments. I’m seeing `http://undefined/embed/bootstrap` in the console on this page, so looks like it’s not getting the rootURL from config.

Fixed, thanks!

> * I also did some searches for “pym” and “iframe” in the code. I’m still seeing a few mentions of both in the code that I think can be removed/updated. For example, in `retrieveSiteFromQuery`, `CallOut.tsx`, `getAriaPoliteMacOSWorkaround`, and a few more files. Also some docs such as in `seo.md`.

Thanks! I removed all of the `pym` reference and most of the `iframe` as well. We still support AMP which puts Coral in an iframe so sometimes it still is valid.

The `seo.md` must probably adapted, or the section about iframe fully dropped. Probably @losowsky & @tessalt know what to do with it.

alinhle committed:
fix: use full url for frame media

alinhle committed:
chore: workaround for half pollyfilled requestIdleCallback functionality...

alinhle committed:
fix: responsive issue

alinhle committed:
fix: styles not applied

alinhle committed:
fix: scrolling

marcushaddon committed:
Merge branch 'develop' into feat/CORL-2299-iframeless-epic

kabeaty committed:
Merge branch 'develop' into feat/CORL-2299-iframeless-epic

kabeaty committed:
fix linting error

kabeaty committed:
Merge branch 'develop' into feat/CORL-2299-iframeless-epic
